File: src/zql/ctx.zig
const std = @import("std");
const types = @import("../types.zig");
const collection = @import("../collection.zig");
const err = @import("errors.zig");
const mem = @import("memory");

pub const ZqlFunc = *const fn (ctx: *ZqlContext, allocator: std.mem.Allocator) anyerror!void;

pub const ZqlContextErr = error {
    FunctionNotSet
};

pub const ZqlTrace = struct {
    value: ?*mem.AutoPtr(types.Value),
    err: ?[]u8,
};

// ZqlContext is used to hold context data for a function
// or exception if happened
pub const ZqlContext = struct {
    db: *collection.Collection,
    args: std.ArrayList(ZqlTrace),
    func: ?ZqlFunc,
    buffer: ?*mem.AutoPtr(types.Value),
    user: ?[]const u8,
    err: ?[]u8,
    
    pub fn init(db: *collection.Collection, allocator: std.mem.Allocator) ZqlContext {
        return ZqlContext{
            .db = db,
            .args = std.ArrayList(ZqlTrace).init(allocator),
            .func = null,
            .buffer = null,
            .err = null,
            .user = null,
        };
    }

    pub fn deinit(self: *ZqlContext) void {
        for (self.args.items) |arg| {
            if (arg.value) |v| {
                v.deinit();
            }
        }
        self.args.deinit();
    }

    pub fn set_user(self: *ZqlContext, username: []const u8) void {
        self.user = username; 
    }

    pub fn get_user(self: *ZqlContext) []const u8 {
        return self.user.?;
    }

    pub fn add_arg(self: *ZqlContext, arg: ZqlTrace) !void {
        try self.args.append(arg);
    }

    pub fn get_arg(self: *const ZqlContext, index: usize) ?*mem.AutoPtr(types.Value) {
        if (index > self.args.items.len) return null;

        return self.args.items[index].value;
    }

    pub fn get_arg_count(self: *const ZqlContext) usize {
        return self.args.items.len;
    }

    pub fn set_fn(self: *ZqlContext, func: ZqlFunc) void {
        self.func = func;
    }

    pub fn execute(self: *ZqlContext, buff: ?*mem.AutoPtr(types.Value), allocator: std.mem.Allocator) !?[]u8 {
        if (buff) |b| {
            self.buffer = b;
        }

        for (self.args.items) |a| {
            if (a.err) |e| {
                return e;
            }
        }

        if (self.func) |func| {
            try func(self, allocator);
            return null;
        }

        return ZqlContextErr.FunctionNotSet;
    }
};

File: src/zql/ctx.cpp
#include <string>
#include <memory>

#include <accounts.hpp>
#include <types.hpp>
#include <zql/ctx.hpp>

namespace ZeonDB::ZQL {
	void Context::set_user(std::string username) {
		this->user = username;
	}

	std::string Context::get_user() {
		return this->user;
	}

	void Context::add_arg(ZqlTrace trace) {
		this->args.push_back(trace);
	}

	std::shared_ptr<ZeonDB::Types::Value> Context::get_arg(size_t index) {
		if (index >= this->args.size()) return nullptr;
		return this->args[index].value;
	}

	void Context::set_options(std::map<std::string, ZeonDB::Types::ManagedValue>* opts) {
		this->options = opts;
	}

	std::map<std::string, ZeonDB::Types::ManagedValue>* Context::get_options() {
		return this->options;
	}

	ZeonDB::Accounts::AccountManager* Context::get_account_manager() {
		return this->amgr;
	}

	ZeonDB::TemplateStore *Context::get_template_store() {
		return this->templ_store;
	}

	std::string Context::get_arg_error(size_t index) {
		if (index >= this->args.size()) return "";
		return this->args[index].error;
	}

	ZeonDB::Accounts::Permission Context::get_perm(std::string key) {
		return this->db->v.c.get_perms(this->user, key);
	}

	std::shared_ptr<ZeonDB::Types::Value> Context::get_db() {
		return this->db;
	}

	size_t Context::arg_count() {
		return this->args.size();
	}

	void Context::set_function(ZqlFunction fn) {
		this->fn = fn;
	}

	void Context::execute(std::shared_ptr<ZeonDB::Types::Value>* buffer) {
		this->temporary_buffer = buffer;

		if (this->fn == nullptr) {
			this->error = "Function not set!";
			return;
		}

		this->fn(this);
	}
}
